File: go-live-stream/internal/handlers/rooms.go
package handlers

import (
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
	"go-live-stream/internal/hub"
	"go-live-stream/internal/models"
)

type RoomHandler struct {
	streamHub *hub.StreamHub
}

func NewRoomHandler(sh *hub.StreamHub) *RoomHandler {
	return &RoomHandler{streamHub: sh}
}

type createRoomRequest struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	HostName    string `json:"hostName"`
}

func (h *RoomHandler) CreateRoom(c echo.Context) error {
	var req createRoomRequest
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request"})
	}
	if req.Title == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "title is required"})
	}
	if req.HostName == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "hostName is required"})
	}

	info := models.RoomInfo{
		ID:          uuid.New().String(),
		Title:       req.Title,
		Description: req.Description,
		HostName:    req.HostName,
		IsLive:      false,
		CreatedAt:   time.Now(),
	}
	h.streamHub.CreateRoom(info)
	return c.JSON(http.StatusCreated, info)
}

func (h *RoomHandler) ListRooms(c echo.Context) error {
	rooms := h.streamHub.ListRooms()
	if rooms == nil {
		rooms = []models.RoomInfo{}
	}
	return c.JSON(http.StatusOK, rooms)
}

func (h *RoomHandler) GetRoom(c echo.Context) error {
	id := c.Param("id")
	info, ok := h.streamHub.GetRoomInfo(id)
	if !ok {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "room not found"})
	}
	return c.JSON(http.StatusOK, info)
}

func (h *RoomHandler) DeleteRoom(c echo.Context) error {
	id := c.Param("id")
	if _, ok := h.streamHub.GetRoomInfo(id); !ok {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "room not found"})
	}
	h.streamHub.DeleteRoom(id)
	return c.NoContent(http.StatusNoContent)
}

File: go-live-stream/frontend/src/api/rooms.js
const BASE = '/api'

export async function listRooms() {
  const res = await fetch(`${BASE}/rooms`)
  if (!res.ok) throw new Error('Failed to fetch rooms')
  return res.json()
}

export async function getRoom(id) {
  const res = await fetch(`${BASE}/rooms/${id}`)
  if (!res.ok) throw new Error('Room not found')
  return res.json()
}

export async function createRoom({ title, description, hostName }) {
  const res = await fetch(`${BASE}/rooms`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ title, description, hostName }),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({}))
    throw new Error(err.error || 'Failed to create room')
  }
  return res.json()
}

export async function deleteRoom(id) {
  const res = await fetch(`${BASE}/rooms/${id}`, { method: 'DELETE' })
  if (!res.ok) throw new Error('Failed to delete room')
}
